src/package.rs

```diff
@@ -13,78 +13,96 @@
 // limitations under the License.
 
 use std::{
-    collections::HashMap,
+    collections::BTreeMap,
+    env::current_dir,
     fmt::{self, Formatter},
-    fs::File,
     io::{self, Cursor, Read, Write},
     ops::Deref,
     path::{Path, PathBuf},
     str::FromStr,
-    sync::Arc,
 };
 
-use async_recursion::async_recursion;
 use bytes::{Buf, Bytes};
-use miette::{ensure, miette, Context, Diagnostic, IntoDiagnostic};
-use semver::{Version, VersionReq};
+use miette::{ensure, miette, Context, IntoDiagnostic};
+use semver::Version;
 use serde::{Deserialize, Serialize};
-use thiserror::Error;
 use tokio::fs;
 use walkdir::WalkDir;
 
 use crate::{
-    credentials::Credentials,
     errors::{DeserializationError, SerializationError},
-    lock::{LockedPackage, Lockfile},
-    manifest::{self, Dependency, Manifest, MANIFEST_FILE},
-    registry::{Artifactory, RegistryUri},
+    lock::LockedPackage,
+    manifest::{self, Manifest, MANIFEST_FILE},
+    registry::RegistryUri,
     ManagedFile,
 };
 
 /// IO abstraction layer over local `buffrs` package store
-pub struct PackageStore;
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PackageStore {
+    root: PathBuf,
+}
 
 impl PackageStore {
     /// Path to the proto directory
-    pub const PROTO_PATH: &str = "proto";
+    const PROTO_PATH: &str = "proto";
     /// Path to the dependency store
-    pub const PROTO_VENDOR_PATH: &str = "proto/vendor";
+    const PROTO_VENDOR_PATH: &str = "proto/vendor";
+
+    fn new(root: PathBuf) -> Self {
+        Self { root }
+    }
+
+    /// Open current directory.
+    pub fn current() -> Result<Self, io::Error> {
+        Ok(Self::new(current_dir()?))
+    }
+
+    /// Open given directory.
+    pub fn open(path: &Path) -> Self {
+        Self::new(path.into())
+    }
+
+    /// Path to the `proto` directory.
+    pub fn proto_path(&self) -> PathBuf {
```

Comment:
Imo this should not return the pathbuf? can we return a borrowed path to prevent an alloc?

Reply:
It returns the `root` path with a directory appended.

I would not really worry about allocations here, the cost is negligible and changing it would be a premature optimization. But we could add an issue for it for someone to tackle at some point.

Perhaps the package store itself could just contain the path to the vendor dir.

Reply:
Its more about semantics than allocations, returning a borrowed path makes the lifetime of the path the same, or shorter, than the store

But also i dont really care